### PR TITLE
Add admin tools page

### LINKED ISSUE #32: Create Tools page
## Description

The plugin should introduce a new page under Tools to allow users to run checks of plugins.

## Acceptance Criteria

- Class `Admin_Page` should exist at `includes/Admin`
- `Admin_Page` class should have the `__construct()` method.
- `add_hooks()` method should exists and have the `add_action( 'admin_menu', array( $this, 'add_page' ) );` hook registering. (**NOTE:** `add_hooks()` method will be called from wherever the class is instantiated).
- The `add_page()` will have `add_management_page()` with below params.
   - `$page_title` should be "Plugin Check"
   - `$menu_title` should be "Plugin Check"
   - `$capability` should be "activate_plugins"
   - `$menu_slug` should be "plugin-check"
   - Define callback method in class `Admin_Page` for `render_page `
   - `$position` should be anything or default
 - The plugin check admin page should be available as below:
 Screenshot 2022-11-21 at 7 56 24 PM
   
- `Admin_Page` class will have the `render_page()`, and it should print the dropdown with the list of plugins and the `check it!` button.
- Use the `get_plugins()` method to get a list of all the activated plugins, exclude `plugin check` from it and show it in the dropdown, like below:

Screenshot 2022-11-21 at 7 58 21 PM

## Tests Coverage

**Automated Tests:**
- Create an `Admin_Page_Test` class that checks whether the `admin_menu` filter is registering properly. 
- Checks that the `Plugin Check` page is registered or not in the `global $submenu`
- Checks the output of the `render_page()` by comparing the output of the plugin's dropdown list.

**Manual Tests:**
- Login to the admin area, and find the `Plugin Check` page available under the `Tools` menu.
- Check whether the dropdown with `the list of plugins` and `check it!` button is available.
- Check on the check it button and check if that is clickable or doing any process.

### PR DESCRIPTION
**Pull request by vishalkakadiya**
Add admin tools page

- Addresses the acceptance criteria in https://github.com/10up/plugin-check/issues/32

Closes https://github.com/10up/plugin-check/issues/32

### CODE REVIEW

File: includes/Admin/Admin_Page.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+/**
+ * Class is handling admin tools page functionality.
+ *
+ * @since n.e.x.t
+ */
+class Admin_Page {
+
+	/**
+	 * Initialize hooks.
```
Comment (felixarntz) [inline includes/Admin/Admin_Page.php]: Please remember that all methods and function descriptions should start with a third-person verb. Also, every method and function should have a `@since n.e.x.t` annotation.
```suggestion
	 * Initializes hooks.
	 *
	 * @since n.e.x.t
```

File: includes/Admin/Admin_Page.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	 */
+	public function add_hooks() {
+		add_action( 'admin_menu', array( $this, 'add_page' ) );
+	}
+
+	/**
+	 * Register the admin page under the tools menu.
```
Comment (felixarntz) [inline includes/Admin/Admin_Page.php]: See above.
```suggestion
	 * Registers the admin page under the tools menu.
	 *
	 * @since n.e.x.t
```

Please update this for the other class methods as well.

File: includes/Admin/Admin_Page.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	 */
+	public function add_page() {
+		add_management_page( __( 'Plugin Check', 'plugin-check' ), __( 'Plugin Check', 'plugin-check' ), 'activate_plugins', 'plugin-check', array( $this, 'render_page' ) );
```
Comment (felixarntz) [inline includes/Admin/Admin_Page.php]: Nit-pick, but maybe break this down into multiple lines? It's a long line and therefore hard to parse right now.

File: includes/Admin/Admin_Page.php [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	}
+
+	/**
+	 * Returns the list of plugins.
+	 *
+	 * @return array List of plugins.
+	 */
+	public function get_available_plugins() {
```
Comment (felixarntz) [inline includes/Admin/Admin_Page.php]: This should not be a public method. It's not called anywhere else, and we should only make methods public if they have to be.

File: templates/admin-page.php [hunk 5]
```diff
@@ -0,0 +1,40 @@
+<?php
+/**
+ * Template for the Admin page.
+ *
+ * @package plugin-check
+ */
+
+if ( empty( $available_plugins ) ) {
+	return;
+}
```
Comment (felixarntz) [inline templates/admin-page.php]: Let's remove this. Imagine a site has no plugins installed. It would be a bad experience showing the user a blank admin page.

File: templates/admin-page.php [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+				<option><?php esc_html_e( 'Select Plugin', 'plugin-check' ); ?></option>
+				<?php foreach ( $available_plugins as $plugin_basename => $available_plugin ) { ?>
+					<option value="<?php echo esc_attr( $plugin_basename ); ?>">
+						<?php echo esc_html( $available_plugin['Name'] ); ?>
+					</option>
+				<?php } ?>
+			</select>
+
+			<input type="submit" value="<?php esc_attr_e( 'Check it!', 'plugin-check' ); ?>" />
+		</form>
```
Comment (felixarntz) [inline templates/admin-page.php]: Instead, I think this code should be wrapped in a check `if ( ! empty( $available_plugins ) )`.

We should then also add an `else` clause that outputs e.g. a paragraph like `No plugins available.`.

File: templates/admin-page.php [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+if ( empty( $available_plugins ) ) {
+	return;
+}
+?>
+
+<div class="wrap">
+
+	<h1><?php esc_html_e( 'Plugin Check', 'plugin-check' ); ?></h1>
+
+	<div class="card">
```
Comment (felixarntz) [inline templates/admin-page.php]: Why display this content in a card?

File: tests/Admin/Admin_Page_Tests.php [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+	public function test_get_plugins() {
+		$available_plugins      = get_plugins();
+		$plugin_check_base_name = plugin_basename( WP_PLUGIN_CHECK_MAIN_FILE );
+
+		if ( isset( $available_plugins[ $plugin_check_base_name ] ) ) {
+			unset( $available_plugins[ $plugin_check_base_name ] );
+		}
+
+		$this->assertEquals( $available_plugins, $this->admin_page->get_available_plugins() );
+	}
```
Comment (felixarntz) [inline tests/Admin/Admin_Page_Tests.php]: Instead of testing this method (which should be private, see above), we should rather test the `render_page()` method for whether it produces the correct output.

**Review by felixarntz — COMMENTED** [review]
@vishalkakadiya Mostly looks good. A few small things to iterate on.

File: includes/Admin/Admin_Page.php (hunk 1 above)
Comment (vishalkakadiya) [reply includes/Admin/Admin_Page.php]: I have updated this now, thank you!

File: includes/Admin/Admin_Page.php (hunk 2 above)
Comment (vishalkakadiya) [reply includes/Admin/Admin_Page.php]: I have updated this now, thank you!

File: includes/Admin/Admin_Page.php (hunk 4 above)
Comment (vishalkakadiya) [reply includes/Admin/Admin_Page.php]: I have updated this now, thank you!

File: includes/Admin/Admin_Page.php (hunk 3 above)
Comment (vishalkakadiya) [reply includes/Admin/Admin_Page.php]: Done

File: templates/admin-page.php (hunk 5 above)
Comment (vishalkakadiya) [reply templates/admin-page.php]: Done

File: templates/admin-page.php (hunk 7 above)
Comment (vishalkakadiya) [reply templates/admin-page.php]: I thought it will look good in card (like below screenshot), but I have removed it now we can circle back on this when we actually implement output of the checks.

Screenshot 2023-02-20 at 12 39 35 PM

File: tests/Admin/Admin_Page_Tests.php (hunk 8 above)
Comment (vishalkakadiya) [reply tests/Admin/Admin_Page_Tests.php]: Done

**Comment by vishalkakadiya** [comment]
@felixarntz I have addressed your feedback. Thank you!

File: templates/admin-page.php (hunk 6 above)
Comment (vishalkakadiya) [reply templates/admin-page.php]: Done

File: tests/Admin/Admin_Page_Tests.php [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+		set_current_screen( $current_screen );
+	}
+
+	public function test_render_page() {
+		$available_plugins      = get_plugins();
+		$plugin_check_base_name = plugin_basename( WP_PLUGIN_CHECK_MAIN_FILE );
+
+		if ( isset( $available_plugins[ $plugin_check_base_name ] ) ) {
+			unset( $available_plugins[ $plugin_check_base_name ] );
+		}
```
Comment (felixarntz) [inline tests/Admin/Admin_Page_Tests.php]: What does this do? Seems unrelated to the test code below.

File: tests/Admin/Admin_Page_Tests.php [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+		wp_set_current_user( $admin_user );
+		set_current_screen( 'dashboard' );
+
+		$this->admin_page->add_page();
+
+		$this->assertArrayHasKey( 'plugin-check', $_parent_pages );
+		$this->assertEquals( 'tools.php', $_parent_pages['plugin-check'] );
+
+		set_current_screen( $current_screen );
```
Comment (felixarntz) [inline tests/Admin/Admin_Page_Tests.php:46]: This does not work reliably - if an assertion above fails, this code doesn't run.

We need to assign the `$_parent_pages` global to a local variable _before_ doing the assertions, then reset `set_current_screen`, and only at the very end run the assertions on the local variable.

File: tests/Admin/Admin_Page_Tests.php [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+		ob_start();
+		$this->admin_page->render_page();
+		$output = ob_get_contents();
+		ob_end_clean();
+
+		$this->assertStringContainsString( 'Check the Plugin', $output );
+		$this->assertStringContainsString( ' id="plugin-check__plugins"', $output );
+		$this->assertStringContainsString( ' name="plugin_check_plugins"', $output );
+		$this->assertStringContainsString( 'Select Plugin', $output );
+		$this->assertStringNotContainsString( ' name="Check it!"', $output );
```
Comment (felixarntz) [inline tests/Admin/Admin_Page_Tests.php]: It would be good to assert here that the `$output` _does not_ include an `<option>` for the `$plugin_check_base_name`, to ensure that the exclusion logic in `Admin_Page::get_available_plugins()` works correctly.

**Review by felixarntz — COMMENTED** [review]
@vishalkakadiya A few last bits of feedback here, but looks almost good to go now.

File: tests/Admin/Admin_Page_Tests.php (hunk 9 above)
Comment (vishalkakadiya) [reply tests/Admin/Admin_Page_Tests.php]: Removed this now, thanks!

File: tests/Admin/Admin_Page_Tests.php (hunk 11 above)
Comment (vishalkakadiya) [reply tests/Admin/Admin_Page_Tests.php]: Done

File: tests/Admin/Admin_Page_Tests.php (hunk 10 above)
Comment (vishalkakadiya) [reply tests/Admin/Admin_Page_Tests.php:46]: Done

**Comment by vishalkakadiya** [comment]
@felixarntz I have addressed the last feedback. Thanks!

**Review by felixarntz — APPROVED** [review]
@vishalkakadiya Thanks, LGTM!

File: tests/Admin/Admin_Page_Tests.php [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+		$parent_pages = $_parent_pages;
+
+		set_current_screen( $current_screen );
+
+		$this->assertArrayHasKey( 'plugin-check', $parent_pages );
+		$this->assertEquals( 'tools.php', $parent_pages['plugin-check'] );
+	}
+
+	public function test_render_page() {
```
Comment (joemcgill) [inline tests/Admin/Admin_Page_Tests.php:52]: This process of unit testing for certain strings when the template is rendered seems somewhat fragile to me, since changes to the template would also require us to update the strings these assertions are looking for. I'd prefer that we mock rendering the template file with some expected data in `$available_plugins`. We can probably override the global `$wp_object_cache` object so the 'plugins' key in the 'plugins' group contains our expected values (since that is what `get_plugins()` uses under the hood) and then confirm that the output we're getting is what we expect.

We should also include a test to ensure that we're getting the correct output whenever no plugins are availble.

**Review by joemcgill — COMMENTED** [review]
This looks pretty good to me besides one suggestion about how we're unit testing the render callback for the admin page.

**Review by jjgrainger — APPROVED** [review]
Thanks @vishalkakadiya looks good!

File: includes/Admin/Admin_Page.php [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+ * Class is handling admin tools page functionality.
+ *
+ * @since n.e.x.t
+ */
+class Admin_Page {
+
+	/**
+	 * Initializes hooks.
+	 */
+	public function add_hooks() {
```
Comment (felixarntz) [inline includes/Admin/Admin_Page.php:22]: This and the other methods in this class are missing `@since n.e.x.t` tags.

**Review by felixarntz — COMMENTED** [review]
@vishalkakadiya While addressing @joemcgill's feedback, can you please also add the missing `@since` annotations?

File: tests/Admin/Admin_Page_Tests.php (hunk 12 above)
Comment (jjgrainger) [reply tests/Admin/Admin_Page_Tests.php:52]: Thanks @joemcgill I've updated the tests to use the object cache to set the expected plugins and tested those. I've also included a test for when there are no plugins available.

File: includes/Admin/Admin_Page.php (hunk 13 above)
Comment (jjgrainger) [reply includes/Admin/Admin_Page.php:22]: Thanks @felixarntz I've updated the PR with the missing `@since` tags.

**Comment by jjgrainger** [comment]
Thanks @felixarntz @joemcgill as the PR has 2 approvals and the remaining feedback has been addressed I'll merge this PR.
